File: newwork-backend/app/models/mcp_server.py
"""
MCP Server model.

Represents a Model Context Protocol server configuration.
"""

from datetime import datetime
from enum import Enum
from typing import Optional, List, Dict, Any
from pydantic import BaseModel, Field


class MCPServerStatus(str, Enum):
    """MCP server connection status."""

    CONNECTED = "connected"
    DISCONNECTED = "disconnected"
    CONNECTING = "connecting"
    ERROR = "error"


class MCPServer(BaseModel):
    """MCP server configuration model."""

    id: str
    name: str
    description: Optional[str] = None
    endpoint: str
    server_type: str = "remote"  # "remote" or "local"
    status: MCPServerStatus = MCPServerStatus.DISCONNECTED
    available_tools: List[str] = Field(default_factory=list)
    capabilities: Optional[Dict[str, Any]] = None
    created_at: datetime = Field(default_factory=datetime.utcnow)
    last_connected_at: Optional[datetime] = None
    config: Optional[Dict[str, Any]] = None  # Additional config (env vars, oauth, etc.)

    class Config:
        use_enum_values = True


class MCPServerCreate(BaseModel):
    """Schema for creating a new MCP server."""

    name: str = Field(..., min_length=1, max_length=100)
    description: Optional[str] = Field(None, max_length=500)
    endpoint: str = Field(..., min_length=1)
    server_type: str = "remote"  # "remote" or "local"
    config: Optional[Dict[str, Any]] = None


class MCPServerUpdate(BaseModel):
    """Schema for updating an MCP server."""

    name: Optional[str] = Field(None, min_length=1, max_length=100)
    description: Optional[str] = Field(None, max_length=500)
    endpoint: Optional[str] = Field(None, min_length=1)
    config: Optional[Dict[str, Any]] = None


class MCPServerResponse(BaseModel):
    """Schema for MCP server API responses."""

    id: str
    name: str
    description: Optional[str] = None
    endpoint: str
    server_type: str
    status: MCPServerStatus
    available_tools: List[str]
    capabilities: Optional[Dict[str, Any]]
    created_at: datetime
    last_connected_at: Optional[datetime]
    config: Optional[Dict[str, Any]]

File: newwork-app/lib/data/models/mcp_server.dart
enum MCPServerStatus { connected, disconnected, connecting, error }

class MCPServer {
  final String id;
  final String name;
  final String? description;
  final String endpoint;
  final MCPServerStatus status;
  final List<String> availableTools;
  final Map<String, dynamic>? capabilities;
  final DateTime createdAt;
  final DateTime? lastConnectedAt;

  MCPServer({
    required this.id,
    required this.name,
    this.description,
    required this.endpoint,
    this.status = MCPServerStatus.disconnected,
    this.availableTools = const [],
    this.capabilities,
    required this.createdAt,
    this.lastConnectedAt,
  });

  factory MCPServer.fromJson(Map<String, dynamic> json) {
    return MCPServer(
      id: json['id'] as String,
      name: json['name'] as String,
      description: json['description'] as String?,
      endpoint: json['endpoint'] as String,
      status: MCPServerStatus.values.firstWhere(
        (e) => e.name == json['status'],
        orElse: () => MCPServerStatus.disconnected,
      ),
      availableTools:
          (json['available_tools'] as List<dynamic>?)
              ?.map((e) => e as String)
              .toList() ??
          [],
      capabilities: json['capabilities'] as Map<String, dynamic>?,
      createdAt: DateTime.parse(json['created_at'] as String),
      lastConnectedAt: json['last_connected_at'] != null
          ? DateTime.parse(json['last_connected_at'] as String)
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'description': description,
      'endpoint': endpoint,
      'status': status.name,
      'available_tools': availableTools,
      'capabilities': capabilities,
      'created_at': createdAt.toIso8601String(),
      'last_connected_at': lastConnectedAt?.toIso8601String(),
    };
  }

  MCPServer copyWith({
    String? id,
    String? name,
    String? description,
    String? endpoint,
    MCPServerStatus? status,
    List<String>? availableTools,
    Map<String, dynamic>? capabilities,
    DateTime? createdAt,
    DateTime? lastConnectedAt,
  }) {
    return MCPServer(
      id: id ?? this.id,
      name: name ?? this.name,
      description: description ?? this.description,
      endpoint: endpoint ?? this.endpoint,
      status: status ?? this.status,
      availableTools: availableTools ?? this.availableTools,
      capabilities: capabilities ?? this.capabilities,
      createdAt: createdAt ?? this.createdAt,
      lastConnectedAt: lastConnectedAt ?? this.lastConnectedAt,
    );
  }

  @override
  String toString() {
    return 'MCPServer(id: $id, name: $name, status: $status, toolCount: ${availableTools.length})';
  }
}


/// MCP Tool model representing a tool available on an MCP server.
class MCPTool {
  final String name;
  final String? description;
  final Map<String, dynamic> inputSchema;

  MCPTool({
    required this.name,
    this.description,
    this.inputSchema = const {},
  });

  factory MCPTool.fromJson(Map<String, dynamic> json) {
    return MCPTool(
      name: json['name'] as String,
      description: json['description'] as String?,
      inputSchema: json['input_schema'] as Map<String, dynamic>? ?? {},
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'name': name,
      'description': description,
      'input_schema': inputSchema,
    };
  }

  @override
  String toString() {
    return 'MCPTool(name: $name, description: $description)';
  }
}


/// MCP Health Status model for health check responses.
class MCPHealthStatus {
  final String serverName;
  final bool isHealthy;
  final String state;
  final double? latencyMs;
  final String? lastPing;
  final String? error;

  MCPHealthStatus({
    required this.serverName,
    required this.isHealthy,
    required this.state,
    this.latencyMs,
    this.lastPing,
    this.error,
  });

  factory MCPHealthStatus.fromJson(Map<String, dynamic> json) {
    return MCPHealthStatus(
      serverName: json['server_name'] as String,
      isHealthy: json['is_healthy'] as bool,
      state: json['state'] as String,
      latencyMs: (json['latency_ms'] as num?)?.toDouble(),
      lastPing: json['last_ping'] as String?,
      error: json['error'] as String?,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'server_name': serverName,
      'is_healthy': isHealthy,
      'state': state,
      'latency_ms': latencyMs,
      'last_ping': lastPing,
      'error': error,
    };
  }

  @override
  String toString() {
    return 'MCPHealthStatus(serverName: $serverName, isHealthy: $isHealthy, latencyMs: $latencyMs)';
  }
}
